group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2904,6 +2908,140 @@ private void validateGenericGroupHeartbeat(
         }
     }
 
+    /**
+     * Handle a generic LeaveGroupRequest.
+     *
+     * @param context        The request context.
+     * @param request        The actual LeaveGroup request.
+     *
+     * @return The LeaveGroup response and the GroupMetadata record to append if the group
+     *         no longer has any members.
+     */
+    public CoordinatorResult<LeaveGroupResponseData, Record> genericGroupLeave(
+        RequestContext context,
+        LeaveGroupRequestData request
+    ) throws UnknownMemberIdException, GroupIdNotFoundException {
+        GenericGroup group = getOrMaybeCreateGenericGroup(request.groupId(), false);
+        if (group.isInState(DEAD)) {
+            return new CoordinatorResult<>(
+                Collections.emptyList(),
+                new LeaveGroupResponseData()
+                    .setErrorCode(COORDINATOR_NOT_AVAILABLE.code())
+            );
+        }
+
+        CoordinatorResult<Void, Record> coordinatorResult = EMPTY_RESULT;
+        List<MemberResponse> memberResponses = new ArrayList<>();
+
+            for (MemberIdentity member: request.members()) {
+                // The LeaveGroup API allows administrative removal of members by GroupInstanceId
+                // in which case we expect the MemberId to be undefined.
+                if (UNKNOWN_MEMBER_ID.equals(member.memberId())) {
+                    if (member.groupInstanceId() != null && group.hasStaticMember(member.groupInstanceId())) {
+                        coordinatorResult = removeCurrentMemberFromGenericGroup(
+                            group,
+                            group.staticMemberId(member.groupInstanceId()),
+                            member.reason()
+                        );
+                        memberResponses.add(
+                            new MemberResponse()
+                                .setMemberId(member.memberId())
+                                .setGroupInstanceId(member.groupInstanceId())
+                        );
+                    } else {
+                        memberResponses.add(
+                            new MemberResponse()
+                                .setMemberId(member.memberId())
+                                .setGroupInstanceId(member.groupInstanceId())
+                                .setErrorCode(Errors.UNKNOWN_MEMBER_ID.code())
+                        );
+                    }
+                } else if (group.isPendingMember(member.memberId())) {
+                    coordinatorResult = removePendingMemberAndUpdateGenericGroup(group, member.memberId());
+                    timer.cancel(genericGroupHeartbeatKey(group.groupId(), member.memberId()));
+                    log.info("Pending member {} has left group {} through explicit `LeaveGroup` request. Reason: {}",
+                        member.memberId(), group.groupId(), member.reason());
```

Comment:
nit: `member.reason()` may be `null` here. We have code in `removeCurrentMemberFromGenericGroup` to handle this case. We should do the same here. The best would be to get that code out of `removeCurrentMemberFromGenericGroup` and sanitize it in this method.